synapseclient/core/version_check.py

```diff
@@ -15,84 +15,176 @@
+    Recommends upgrade, if a newer version exists.
 
-        levels = 3 if check_for_point_releases else 2
+    This has been split of from the version_check function to make testing easier.
+
+    Args:
+        current_version (Optional[str], optional): The current version of the package.
+          Defaults to None.
+          This argument is mainly used for testing.
+        check_for_point_releases (bool, optional):
+          Defaults to False.
+          If True, The whole package versions will be compared (ie. 1.0.0)
+          If False, only the major and minor package version will be compared (ie. 1.0)
+        use_local_metadata (bool, optional):
+          Defaults to False.
+          If True, importlib.resources will be used to get the latest version fo the package
+          If False, the latest version fo the package will be taken from Pypi
 
-        # Compare with latest version
-        if _version_tuple(current_version, levels=levels) < _version_tuple(
-            version_info["latestVersion"], levels=levels
-        ):
-            sys.stderr.write(
-                "\nUPGRADE AVAILABLE\n\nA more recent version of the Synapse Client"
-                " (%s) is available. Your version (%s) can be upgraded by typing:\n   "
-                " pip install --upgrade synapseclient\n\n"
-                % (
-                    version_info["latestVersion"],
-                    current_version,
-                )
-            )
-            if "releaseNotes" in version_info:
-                sys.stderr.write(
-                    "Python Synapse Client version %s release notes\n\n"
-                    % version_info["latestVersion"]
-                )
-                sys.stderr.write(version_info["releaseNotes"] + "\n\n")
-            return False
+    Returns:
+        bool: True if current version is the latest release (or higher) version, otherwise False.
+    """
+    if not current_version:
+        current_version = synapseclient.__version__
+    assert isinstance(current_version, str)
+
+    if use_local_metadata:
+        metadata = _get_local_package_metadata()
+        latest_version = metadata["latestVersion"]
+        assert isinstance(latest_version, str)
+    else:
+        latest_version = _get_version_info_from_pypi()
 
-    except Exception as e:
-        # Don't prevent the client from running if something goes wrong
-        sys.stderr.write("Exception in version check: %s\n" % (str(e),))
-        return False
+    levels = 3 if check_for_point_releases else 2
 
+    if _is_current_version_behind(current_version, latest_version, levels):
+        _write_package_behind_messages(current_version, latest_version)
+        return False
     return True
 
 
+def _get_version_info_from_pypi() -> str:
+    """Gets the current release version from PyPi
+
+    Returns:
+        str: The current release version
+    """
+    with urllib.request.urlopen(_PYPI_JSON_URL) as url:
```

Comment:
Swap this over to use the `httpx` library instead. I'd like for everything to funnel through that library.

https://www.python-httpx.org/quickstart/

Reply:
Done